core/includes/backup/backup.database.inc

```diff
@@ -0,0 +1,350 @@
+    }
+    $file->clearLog();
+
+    return $lines_executed;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function postRestore(BackupFile &$file) {
+    parent::postRestore($file);
+
+    // If a compressed copy of this backup exists, delete the uncompressed copy.
+    if ($file->lastExtension() === 'sql') {
+      $uncompressed_path = $file->filePath();
+      $file->pushExtension('gz');
+      if (file_exists($file->filepath())) {
+        file_unmanaged_delete($uncompressed_path);
+      }
+      else {
+        $file->popExtension();
+      }
+    }
+  }
+
+  /**
+   * Get the db connection for the specified db.
+   */
+  protected function getDatabaseConnection() {
+    if (!$this->connection) {
+      $parts = explode(':', $this->getTarget());
+      // One of the predefined databases (set in settings.php)
+      if ($parts[0] == 'db') {
+        $key = empty($parts[1]) ? 'default' : $parts[1];
+        $target = empty($parts[2]) ? 'default' : $parts[2];
+      }
+      // An arbitrary database URL such as mysql://user:pass@localhost:3306/db.
+      // This ability is not used directly by core, which only provides UIs for
+      // backing up and restoring existing database connections.
+      else {
+        $url_parts = parse_url($this->getTarget());
+        // If the url is specified build it into a connection info array.
+        if ($url_parts) {
+          $info = array(
+            'driver' => empty($url_parts['scheme']) ? NULL : $url_parts['scheme'],
+            'host' => empty($url_parts['host']) ? NULL : $url_parts['host'],
+            'port' => empty($url_parts['port']) ? 3306 : $url_parts['port'],
+            'username' => empty($url_parts['user']) ? NULL : $url_parts['user'],
+            'password' => empty($url_parts['pass']) ? NULL : $url_parts['pass'],
+            'database' => empty($url_parts['path']) ? NULL : $url_parts['path'],
+          );
+          $key = uniqid('backdrop_backup_tmp_');
+          $target = 'default';
+          Database::addConnectionInfo($key, $target, $info);
+        }
+        // No database selected. Assume the default.
+        else {
+          $key = $target = 'default';
+        }
+      }
+      if ($target && $key) {
+        $this->connection = Database::getConnection($target, $key);
+      }
+    }
+    return $this->connection;
+  }
+
+  /**
+   * Backup the databases to a file.
+   *
+   * @return bool
+   *   TRUE is backup successful, FALSE otherwise.
+   */
+  abstract protected function backupDatabaseToFile(BackupFile $file);
+
+  /**
+   * Backup the databases to a file.
+   *
+   * @return int|false
+   *   The number of commands executed on restore. FALSE on failure.
+   */
+  abstract protected function restoreDatabaseFromFile(BackupFile $file);
+
+  /**
+   * Get a list of objects in the database.
+   */
+  protected function getObjectNames() {
+    // Must be overridden.
+    $out = $this->getTableNames();
+    if (method_exists($this, 'getViewNames')) {
```

Comment:
Not sure, if I'm getting this right, but this method will always exist - it's defined a couple of lines below. So what's the purpose of this method_exists()? Or should the check be for something different?

Reply:
You're right. In Backup and Migrate module, these methods were not declared as `abstract` and so it was possible for a child class to not implement it. In this version, child classes must implement it or get a PHP error, so we can remove this check.